File: ts/src/gitstore.ts
/**
 * GitStore: versioned filesystem backed by a bare git repository.
 */

import * as nodeFs from 'node:fs';
import git from 'isomorphic-git';
import type { FsModule, Signature, MirrorDiff, HttpClient } from './types.js';
import { RefDict } from './refdict.js';
import { NoteDict } from './notes.js';

/**
 * A versioned filesystem backed by a bare git repository.
 *
 * Open or create a store with `GitStore.open()`. Access snapshots via
 * `branches`, `tags`, and `notes`.
 */
export class GitStore {
  /** @internal */ _fsModule: FsModule;
  /** @internal */ _gitdir: string;
  /** @internal */ _signature: Signature;

  /** Dict-like access to branches. */
  branches: RefDict;
  /** Dict-like access to tags. */
  tags: RefDict;
  /** Git notes namespaces. */
  notes: NoteDict;

  constructor(fsModule: FsModule, gitdir: string, author: string, email: string) {
    this._fsModule = fsModule;
    this._gitdir = gitdir;
    this._signature = { name: author, email };
    this.branches = new RefDict(this, 'refs/heads/');
    this.tags = new RefDict(this, 'refs/tags/');
    this.notes = new NoteDict(this);
  }

  toString(): string {
    return `GitStore('${this._gitdir}')`;
  }

  /**
   * Open or create a bare git repository.
   *
   * @param path - Path to the bare repository directory.
   * @param opts.fs - Filesystem module (default: Node.js `node:fs`). Override for custom implementations.
   * @param opts.create - Create the repo if it doesn't exist (default: true).
   * @param opts.branch - Initial branch when creating (default: "main"). Null for no branch.
   * @param opts.author - Default author name (default: "vost").
   * @param opts.email - Default author email (default: "vost@localhost").
   */
  static async open(
    path: string,
    opts: {
      fs?: FsModule;
      create?: boolean;
      branch?: string | null;
      author?: string;
      email?: string;
    } = {},
  ): Promise<GitStore> {
    const fsModule = opts.fs ?? nodeFs as unknown as FsModule;
    const create = opts.create ?? true;
    const branch = opts.branch !== undefined ? opts.branch : 'main';
    const author = opts.author ?? 'vost';
    const email = opts.email ?? 'vost@localhost';

    // Check if repo exists
    let exists = false;
    try {
      await fsModule.promises.stat(`${path}/HEAD`);
      exists = true;
    } catch { /* not found */ }

    if (exists) {
      return new GitStore(fsModule, path, author, email);
    }

    if (!create) {
      throw new Error(`Repository not found: ${path}`);
    }

    // Create bare repo
    await git.init({ fs: fsModule, gitdir: path, bare: true });

    const store = new GitStore(fsModule, path, author, email);

    if (branch !== null) {
      // Create initial empty commit on the branch
      const emptyTreeOid = await git.writeTree({ fs: fsModule, gitdir: path, tree: [] });
      const now = Math.floor(Date.now() / 1000);
      const commitOid = await git.writeCommit({
        fs: fsModule,
        gitdir: path,
        commit: {
          message: `Initialize ${branch}\n`,
          tree: emptyTreeOid,
          parent: [],
          author: { name: author, email, timestamp: now, timezoneOffset: 0 },
          committer: { name: author, email, timestamp: now, timezoneOffset: 0 },
        },
      });

      // Create the branch ref
      await git.writeRef({
        fs: fsModule,
        gitdir: path,
        ref: `refs/heads/${branch}`,
        value: commitOid,
      });

      // Set HEAD to point at the branch
      await git.writeRef({
        fs: fsModule,
        gitdir: path,
        ref: 'HEAD',
        value: `refs/heads/${branch}`,
        symbolic: true,
        force: true,
      });
    }

    return store;
  }

  /**
   * Push all refs to url, creating an exact mirror.
   *
   * Remote-only refs are deleted (unless `refs` filtering is used).
   * Supports HTTP URLs, local bare-repo paths, and `.bundle` files.
   *
   * @param url - Remote repository URL, local path, or bundle file path.
   * @param opts.http - HTTP client (required for HTTP URLs only).
   * @param opts.dryRun - Compute diff without pushing.
   * @param opts.onAuth - Optional authentication callback.
   * @param opts.refs - Only backup these refs (short names resolved automatically).
   * @param opts.format - Force format: `'bundle'` for git bundle output.
   * @returns A MirrorDiff describing what changed (or would change).
   */
  async backup(
    url: string,
    opts: {
      http?: HttpClient;
      dryRun?: boolean;
      onAuth?: Function;
      refs?: string[];
      format?: string;
    } = {},
  ): Promise<MirrorDiff> {
    const { backup } = await import('./mirror.js');
    return backup(this, url, opts);
  }

  /**
   * Fetch refs from url additively into the local store.
   *
   * Local-only refs are preserved (not deleted).  All branches, tags,
   * and notes from the source are merged in, but HEAD (the current
   * branch pointer) is not changed — use
   * `store.branches.setCurrent("name")` afterwards if needed.
   *
   * Supports HTTP URLs, local bare-repo paths, and `.bundle` files.
   *
   * @param url - Remote repository URL, local path, or bundle file path.
   * @param opts.http - HTTP client (required for HTTP URLs only).
   * @param opts.dryRun - Compute diff without fetching.
   * @param opts.onAuth - Optional authentication callback.
   * @param opts.refs - Only restore these refs (short names resolved automatically).
   * @param opts.format - Force format: `'bundle'` for git bundle input.
   * @returns A MirrorDiff describing what changed (or would change).
   */
  async restore(
    url: string,
    opts: {
      http?: HttpClient;
      dryRun?: boolean;
      onAuth?: Function;
      refs?: string[];
      format?: string;
    } = {},
  ): Promise<MirrorDiff> {
    const { restore } = await import('./mirror.js');
    return restore(this, url, opts);
  }

  /**
   * Export refs to a bundle file.
   *
   * @param path - Destination `.bundle` file path.
   * @param opts.refs - Only export these refs (short names resolved automatically).
   */
  async bundleExport(
    path: string,
    opts: { refs?: string[] } = {},
  ): Promise<void> {
    const { bundleExport } = await import('./mirror.js');
    return bundleExport(this, path, opts.refs);
  }

  /**
   * Import refs from a bundle file (additive — no deletes).
   *
   * @param path - Source `.bundle` file path.
   * @param opts.refs - Only import these refs (short names resolved automatically).
   */
  async bundleImport(
    path: string,
    opts: { refs?: string[] } = {},
  ): Promise<void> {
    const { bundleImport } = await import('./mirror.js');
    return bundleImport(this, path, opts.refs);
  }
}

File: kotlin/src/main/kotlin/vost/GitStore.kt
package vost

import org.eclipse.jgit.lib.Constants
import org.eclipse.jgit.lib.ObjectId
import org.eclipse.jgit.lib.Repository
import org.eclipse.jgit.lib.StoredConfig
import org.eclipse.jgit.storage.file.FileRepositoryBuilder
import java.io.File

/**
 * A versioned filesystem backed by a bare git repository.
 *
 * Open or create a store with [open]. Access snapshots via
 * [branches], [tags], and [notes].
 */
class GitStore private constructor(
    internal val repo: Repository,
    internal val signature: Signature,
) : AutoCloseable {

    /** Dict-like access to branches. */
    val branches = RefDict(this, "refs/heads/", isTags = false)

    /** Dict-like access to tags. */
    val tags = RefDict(this, "refs/tags/", isTags = true)

    /** Git notes namespaces. */
    val notes = NoteDict(this)

    /**
     * Get an Fs snapshot for any ref (branch, tag, or commit hash).
     *
     * Resolution order: branches -> tags -> commit hash.
     * Writable for branches, read-only for tags and hashes.
     *
     * @param ref Branch name, tag name, or commit hash.
     * @param back Walk back N ancestor commits (default 0).
     * @return Fs snapshot for the resolved ref.
     * @throws NoSuchElementException If the ref cannot be resolved.
     */
    fun fs(ref: String, back: Int = 0): Fs {
        val result = when {
            ref in branches -> branches[ref]
            ref in tags -> tags[ref]
            else -> {
                // Try as commit hash
                val objectId = try {
                    ObjectId.fromString(ref)
                } catch (e: Exception) {
                    throw NoSuchElementException("ref not found: '$ref'")
                }
                val revWalk = org.eclipse.jgit.revwalk.RevWalk(repo)
                try {
                    try {
                        revWalk.parseCommit(objectId)
                    } catch (e: Exception) {
                        throw NoSuchElementException("ref not found: '$ref'")
                    }
                    Fs(this, objectId, writable = false)
                } finally {
                    revWalk.close()
                }
            }
        }
        return if (back > 0) result.back(back) else result
    }

    /**
     * Read raw blob data by hash, bypassing tree/ref resolution.
     * Direct object store lookup — the fastest path to blob data.
     *
     * @param hash 40-char hex SHA of the blob.
     * @param offset Byte offset to start reading from.
     * @param size Maximum number of bytes to return (null for all).
     */
    fun readByHash(hash: String, offset: Int = 0, size: Int? = null): ByteArray {
        val oid = ObjectId.fromString(hash)
        val loader = repo.open(oid, Constants.OBJ_BLOB)
        val data = loader.bytes
        if (offset > 0 || size != null) {
            val start = minOf(offset, data.size)
            val end = if (size != null) minOf(start.toLong() + size.toLong(), data.size.toLong()).toInt() else data.size
            return data.copyOfRange(start, end)
        }
        return data
    }

    /**
     * Check if a blob with the given hash exists in the object store.
     *
     * @param hash 40-char hex SHA of the blob.
     * @return true if the blob exists, false otherwise.
     */
    fun hasHash(hash: String): Boolean {
        return try {
            readByHash(hash, size = 0)
            true
        } catch (e: Exception) {
            false
        }
    }

    override fun toString(): String = "GitStore(${repo.directory})"

    override fun close() {
        repo.close()
    }

    /**
     * Push refs to url (or write a bundle file).
     *
     * Without [refs] this is a full mirror: remote-only refs are deleted.
     * With [refs] only the specified refs are pushed (no deletes).
     * With [refMap] refs are renamed during push (keys=source, values=dest);
     * takes precedence over [refs].
     *
     * @param url Destination URL (local path or remote), or bundle file path.
     * @param dryRun If true, compute diff but don't push.
     * @param refs Optional list of ref names to limit the backup to.
     * @param refMap Optional mapping of source ref names to dest ref names.
     * @param format Optional format string; "bundle" forces bundle output.
     * @return MirrorDiff describing what changed (or would change).
     */
    fun backup(
        url: String,
        dryRun: Boolean = false,
        refs: List<String>? = null,
        refMap: Map<String, String>? = null,
        format: String? = null,
        squash: Boolean = false,
    ): MirrorDiff = MirrorOps.backup(this, url, dryRun, refs, refMap, format, squash)

    /**
     * Fetch refs from url (or import a bundle file).
     *
     * Restore is **additive**: it adds and updates refs but never deletes
     * local-only refs.  HEAD (the current branch pointer) is not changed —
     * use `store.branches.setCurrent("name")` afterwards if needed.
     * With [refMap] refs are renamed when written locally (keys=source,
     * values=dest); takes precedence over [refs].
     *
     * @param url Source URL (local path or remote), or bundle file path.
     * @param dryRun If true, compute diff but don't fetch.
     * @param refs Optional list of ref names to limit the restore to.
     * @param refMap Optional mapping of source ref names to dest ref names.
     * @param format Optional format string; "bundle" forces bundle input.
     * @return MirrorDiff describing what changed (or would change).
     */
    fun restore(
        url: String,
        dryRun: Boolean = false,
        refs: List<String>? = null,
        refMap: Map<String, String>? = null,
        format: String? = null,
    ): MirrorDiff = MirrorOps.restore(this, url, dryRun, refs, refMap, format)

    /**
     * Pack loose objects into a packfile.
     *
     * Consolidates loose git objects into a single packfile for better
     * performance and disk usage.  This implementation uses JGit's GC
     * repack which may also perform additional housekeeping.
     *
     * @return Number of objects packed (as reported by the repacker).
     */
    fun pack(): Int {
        val fileRepo = repo as org.eclipse.jgit.internal.storage.file.FileRepository
        val jgitGc = org.eclipse.jgit.internal.storage.file.GC(fileRepo)
        jgitGc.setExpireAgeMillis(Long.MAX_VALUE)  // don't prune
        jgitGc.setPackExpireAgeMillis(Long.MAX_VALUE)
        val looseCount = jgitGc.getStatistics().numberOfLooseObjects
        jgitGc.repack()
        jgitGc.prunePacked()
        return looseCount.toInt()
    }

    /**
     * Run garbage collection: clean up and pack loose objects.
     *
     * Uses JGit's built-in GC which packs loose objects and may
     * prune unreachable objects.
     *
     * @return Number of objects packed.
     */
    fun gc(): Int {
        val fileRepo = repo as org.eclipse.jgit.internal.storage.file.FileRepository
        val jgitGc = org.eclipse.jgit.internal.storage.file.GC(fileRepo)
        val looseCount = jgitGc.getStatistics().numberOfLooseObjects
        jgitGc.gc().get()  // blocking call
        return looseCount.toInt()
    }

    companion object {
        /**
         * Open or create a bare git repository.
         *
         * @param path Path to the bare repository.
         * @param create If true (default), create the repo when it doesn't exist.
         * @param branch Initial branch name when creating (default "main").
         *               Null to create a bare repo with no branches.
         * @param author Default author name for commits.
         * @param email Default author email for commits.
         * @param compression Zlib compression level for git objects (0-9). Null uses the git default.
         * @param bigFileThreshold Blobs larger than this (bytes) skip delta compression. 0 = all blobs skip deltas. Null uses the git default.
         */
        fun open(
            path: String,
            create: Boolean = true,
            branch: String? = "main",
            author: String = "vost",
            email: String = "vost@localhost",
            compression: Int? = null,
            bigFileThreshold: Long? = null,
        ): GitStore {
            val dir = File(path)

            if (dir.exists()) {
                val repo = FileRepositoryBuilder()
                    .setBare()
                    .setGitDir(dir)
                    .build()
                configureForBareRepo(repo)
                applyConfig(repo, compression, bigFileThreshold)
                return GitStore(repo, Signature(author, email))
            }

            if (!create) {
                throw java.io.FileNotFoundException("Repository not found: $path")
            }

            // Create new bare repo
            dir.mkdirs()
            val repo = FileRepositoryBuilder()
                .setBare()
                .setGitDir(dir)
                .build()
            repo.create(true)
            configureForBareRepo(repo)
            applyConfig(repo, compression, bigFileThreshold)

            val store = GitStore(repo, Signature(author, email))

            if (branch != null) {
                val sig = store.signature
                val inserter = repo.newObjectInserter()
                try {
                    // Create empty tree
                    val emptyTreeId = inserter.insert(org.eclipse.jgit.lib.TreeFormatter())

                    // Create initial commit
                    val commit = org.eclipse.jgit.lib.CommitBuilder()
                    commit.setTreeId(emptyTreeId)
                    commit.setAuthor(org.eclipse.jgit.lib.PersonIdent(sig.name, sig.email))
                    commit.setCommitter(commit.author)
                    commit.setMessage("Initialize $branch\n")

                    val commitId = inserter.insert(commit)
                    inserter.flush()

                    // Set the branch ref
                    val refUpdate = repo.updateRef("refs/heads/$branch")
                    refUpdate.setNewObjectId(commitId)
                    refUpdate.setExpectedOldObjectId(ObjectId.zeroId())
                    refUpdate.setRefLogMessage("commit: Initialize $branch", false)
                    refUpdate.update()

                    // Set HEAD to point to this branch
                    val headUpdate = repo.updateRef(Constants.HEAD)
                    headUpdate.link("refs/heads/$branch")
                } finally {
                    inserter.close()
                }
            }

            return store
        }

        /** Configure bare repo to write reflogs (needed for undo/redo). */
        private fun configureForBareRepo(repo: Repository) {
            val config: StoredConfig = repo.config
            config.setString("core", null, "logAllRefUpdates", "always")
            config.save()
        }

        /** Apply compression and bigFileThreshold config to the repository. */
        private fun applyConfig(repo: Repository, compression: Int?, bigFileThreshold: Long?) {
            if (compression == null && bigFileThreshold == null) return
            val config: StoredConfig = repo.config
            if (compression != null) {
                config.setInt("core", null, "compression", compression)
            }
            if (bigFileThreshold != null) {
                config.setLong("core", null, "bigFileThreshold", bigFileThreshold)
            }
            config.save()
        }
    }
}
